bpo-46968: Query altstack size dynamically on Linux kernel >=5.14

oleksandr-pavlyk opened the pull request:
In Linux kernel 5.14 one can dynamically request size of altstacksize
needed to handle signals based on hardware capabilities. 
This patch sets stack size to be the maximum of current value of SIGSTKSZ * 2 and
SIGSTKSZ + getauxval(AT_MINSIGSTKSZ).

This changes allows for Python extension's request to Linux kernel
to use AMX_TILE instruction set on Sapphire Rapids Xeon processor
to succeed, unblocking use of the ISA in frameworks.

This change proposes a solution to the problem described 
in https://bugs.python.org/msg414809



<!-- issue-number: [bpo-46968](https://bugs.python.org/issue46968) -->
https://bugs.python.org/issue46968
<!-- /issue-number -->

oleksandr-pavlyk committed:
Query altstack size dynamically on Linux kernel >=5.14

In Linux kernel 5.14 one can dynamically request size of altstacksize
based on hardware capabilities. This patch sets stack size to be
the maximum of current value of SIGSTKSZ * 2 and
SIGSTKSZ + getauxval(AT_MINSIGSTKSZ)

This changes allows for Python extension's request to Linux kernel
to use AMX_TILE instruction set on Sapphire Rapids Xeon processor
to succeed, unblocking use of the ISA in frameworks.

gpshead commented on aclocal.m4:
```diff
@@ -1,6 +1,6 @@
-# generated automatically by aclocal 1.16.3 -*- Autoconf -*-
```
We need to use the same versions when regenerating these. I believe the version mismatch is why the generated configure script deletes some unrelated runstatedir lines as well.

https://devguide.python.org/setup/#regenerate-configure

gpshead reviewed (changes requested):
The core of the changes look good, the annoying part is in the generated configure and aclocal.m4 files - those should be generated with the same version as the existing ones.

See the 2022 comment on https://bugs.python.org/issue44035 for a pointer to a docker image that can be used for the autoconf regeneration stuff.  (that image is not yet documented AFAICT)

tiran commented on configure.ac:
```diff
@@ -2405,6 +2405,8 @@ AC_CHECK_HEADERS([net/if.h], [], [],
 #endif
 ])
 
+AC_CHECK_HEADERS([linux/auxvec.h], [], [], [])
```
Move this header check to the large ``AC_CHECK_HEADERS`` around line 2376.

oleksandr-pavlyk committed:
Added check for linux/auxvec.h in configure

Introduced HAVE_LINUX_AUXVEC_H in configure.ac and pyconfig.h.in
Used cpython_autoconf:269 docker container to generate configure

tiran reviewed (changes requested):
You need a verify specific combination of autoconf 2.69 + runstatedir patch, autoarchive version, and pkg-config m4 macro version. Your version of autoconf aclocal macros is too old. If you cannot use the container image to regenerate the files, then apply the changes manually.

* revert ``aclocal.m4`` and ``configure`` changes.
* manually add ``linux/auxvec.h`` to ``configure``

vstinner commented on Modules/faulthandler.c:
```diff
@@ -1372,7 +1379,18 @@ _PyFaulthandler_Init(int enable)
+    }
+#else
     stack.ss_size = SIGSTKSZ * 2;
+#endif
```
Please add a comment to explain why you use AT_MINSIGSTKSZ. I suggest to change the code like that:

```suggestion
    stack.ss_size = SIGSTKSZ * 2;

#ifdef AT_MINSIGSTKSZ
    // bpo-46968: (...)
    unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
    if (at_minstack_size != 0) {
        stack.ss_size = SIGSTKSZ + at_minstack_size;
    }
#endif
```

oleksandr-pavlyk replied:
Done. Thank you for the review.

vstinner commented on Misc/NEWS.d/next/Core and Builtins/2022-03-09-20-10-34.bpo-46968.Rz-uPr.rst:
```diff
@@ -0,0 +1 @@
+Dynamically determine size of sigaltstack CPython allocates to ensure Linux signal delivery on the hardware running the interpreter. The required Linux kernel support is available since kernel version 5.14.
```
Please mention faulthandler somewhere.

oleksandr-pavlyk replied:
Done.

vstinner commented on Modules/faulthandler.c:
```diff
@@ -20,6 +20,19 @@
+#  include <linux/auxvec.h>
+#endif
+#include <sys/auxv.h>
+#endif
```
It's unclear to me why <sys/auxv.h> is always included, but <linux/auxvec.h> is included conditionally. Would it be possible that AT_MINSIGSTKSZ is defined by <sys/auxv.h>?

```suggestion
#ifdef FAULTHANDLER_USE_ALT_STACK
#  ifdef HAVE_LINUX_AUXVEC_H
#    include <linux/auxvec.h>
#  endif
#  include <sys/auxv.h>
#endif
```

oleksandr-pavlyk replied:
`linux/auxvec.h` does not exist on OSX, but without it `getauxval` (include from `auxv.h`) is not used, so I will move the `#include <sys/auxv.h>` inside the `HAVE_LINUX_AUXVEC_H` conditional.

oleksandr-pavlyk replied:
Changes applied.

vstinner reviewed (commented):
I like AT_MINSIGSTKSZ feature! It's better than coarse ``SIGSTKSZ * 2``. Here is a first review.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Misc/NEWS.d/next/Library/2022-03-10-14-51-11.bpo-46968.ym2QxL.rst:
```diff
@@ -0,0 +1 @@
+Dynamically determine size of sigaltstack CPython allocates during initialization of the faulthandler extension to ensure Linux signal delivery on the hardware running the interpreter. The required Linux kernel support is available since kernel version 5.14.
```
I would prefer to put faulthandler at the beginning. Like:

```
:mod:`faulthandler`: On Linux 5.14 and newer, dynamically determine size (...) using ``getauxval(AT_MINSIGSTKSZ)`` (...).
```

Rather than "sigaltstack", I suggest to say: "signal handler stack size".

IMO it's worth it to go into the details:

"This changes allows for Python extension's request to Linux kernel
to use AMX_TILE instruction set on Sapphire Rapids Xeon processor
to succeed, unblocking use of the ISA in frameworks."

nitpick: Can you please try to limit to 80 columns?

oleksandr-pavlyk replied:
I modified the .rst file directly as requested.

vstinner commented on Modules/faulthandler.c:
```diff
@@ -1373,6 +1378,17 @@ _PyFaulthandler_Init(int enable)
        platforms. */
     stack.ss_size = SIGSTKSZ * 2;
+#ifdef AT_MINSIGSTKSZ
+    {
```
These braces are not needed. Python 3.11 uses C11, see: https://peps.python.org/pep-0007/#c-dialect

oleksandr-pavlyk replied:
I had a question here. As is the change will only be effective on the new hardware if CPython was compiled against kernel 5.14 of later. Would be acceptable to do

```c
#ifndef AT_MINSIGSTKSZ
#define AT_MINSIGSTKSZ 51
#endif
```

relying on the fact that `getauxval` returns 0 for unknown tokens?

This would permit backport of this fix to earlier version of CPython, right?

oleksandr-pavlyk replied:
> These braces are not needed. Python 3.11 uses C11, see: https://peps.python.org/pep-0007/#c-dialect

Removed, and indentation adjusted. Would appreciate feedback about tweaking the proposed fix so that the call to `getauxval` is not in the preprocessor branch by defining `AT_MINSIGSTKSZ` if undefined.

vstinner replied:
Please don't hardcode constants in Python: don't add ``#define AT_MINSIGSTKSZ 51``.

oleksandr-pavlyk replied:
> Please don't hardcode constants in Python: don't add `#define AT_MINSIGSTKSZ 51`.

Ok, understood, but had to ask.

vstinner commented on Modules/faulthandler.c:
```diff
@@ -1373,6 +1378,17 @@ _PyFaulthandler_Init(int enable)
+        unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
+        if (at_minstack_size != 0) {
+            stack.ss_size = SIGSTKSZ + at_minstack_size;
+	}
```
strange indentation :-)

oleksandr-pavlyk replied:
Addressed.

vstinner commented on Modules/faulthandler.c:
```diff
@@ -1373,6 +1378,15 @@ _PyFaulthandler_Init(int enable)
+       Linux kernel version >= 5.14 */
+    unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
+    if (at_minstack_size != 0) {
+        stack.ss_size = SIGSTKSZ + at_minstack_size;
```
I'm surprised that AT_MINSIGSTKSZ doesn't return directly the value which should be passed to sigaltstack(), but it seems like your code is the correct way to use the value.

oleksandr-pavlyk replied:
The `AT_MINSIGSTKSZ` is a preprocessor variable. The `getauxval` checks hardware capabilities (e.g. total size of CPU state to be saved at context switches) and may return different values on different platforms.

vstinner replied:
I mean that I expected to write (simplified code without error handling): ``stack.ss_size = getauxval(AT_MINSIGSTKSZ)``, rather than ``stack.ss_size = SIGSTKSZ + getauxval(AT_MINSIGSTKSZ)``. But your code looks correct according to the doc that I found on the Internet. Like the official Linux documentation:

```
ss.ss_size = getauxval(AT_MINSIGSTKSZ) + SIGSTKSZ;
```

ref: https://docs.kernel.org/x86/elf_auxvec.html#introduction

vstinner reviewed (approved):
LGTM.
